{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "e67aba9e-c44a-4cfe-bad5-d3f3e197f05a",
   "metadata": {},
   "outputs": [],
   "source": [
    "from __future__ import print_function\n",
    "import json\n",
    "from matplotlib import pyplot as plt\n",
    "import os\n",
    "import sys\n",
    "import glob\n",
    "from time import time\n",
    "import itertools\n",
    "from itertools import islice, chain\n",
    "from struct import *\n",
    "import shutil\n",
    "import pandas as pd\n",
    "import argparse\n",
    "import errno\n",
    "import math\n",
    "import subprocess\n",
    "\n",
    "# import pickle\n",
    "import dill as pickle \n",
    "import gffutils\n",
    "import pysam\n",
    "import numpy as np\n",
    "import multiprocessing as mp\n",
    "from multiprocessing import Pool\n",
    "from collections import defaultdict\n",
    "# from collections import OrderedDict\n",
    "\n",
    "# from modules import create_splice_graph as splice_graph\n",
    "# from modules import graph_chainer \n",
    "\n",
    "from uLTRA.modules import create_augmented_gene as augmented_gene \n",
    "from uLTRA.modules import mem_wrapper \n",
    "from uLTRA.modules import colinear_solver \n",
    "from uLTRA.modules import help_functions\n",
    "from uLTRA.modules import classify_read_with_mams\n",
    "from uLTRA.modules import classify_alignment2\n",
    "from uLTRA.modules import sam_output\n",
    "from uLTRA.modules import align\n",
    "from uLTRA.modules import prefilter_genomic_reads\n",
    "\n",
    "from new_modules import functions\n",
    "from new_modules import evaluate_exons\n",
    "from new_modules import evaluate_splice_annotations\n",
    "from new_modules import get_diff_loc_reads\n",
    "from uLTRA.evaluation import plot_correctness_per_exon_size\n",
    "from uLTRA.evaluation import plots\n",
    "from uLTRA.evaluation import venn_diagram"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "45112657-344a-48e9-b71b-d542a29be327",
   "metadata": {},
   "outputs": [],
   "source": [
    "from new_modules.arguments import arguments\n",
    "\n",
    "def initialize_dump(outfolder):\n",
    "    if os.path.exists(outfolder) and os.path.isdir(outfolder):\n",
    "        shutil.rmtree(outfolder)\n",
    "    help_functions.mkdir_p(outfolder)\n",
    "    \n",
    "def create_args(ref, gtf, bed, reads, ont, isoseq, disable_infer, name, tool_name, mm2, desalt_index):\n",
    "    args = arguments(ref, gtf, bed, reads, ont, isoseq, disable_infer, name, tool_name, mm2, desalt_index)\n",
    "    s = args.reads\n",
    "    s = s[s.rindex('/')+1:]\n",
    "    s = s[:s.rindex('.')]\n",
    "    args.name = s\n",
    "    if args.ont:\n",
    "        args.min_mem = 17\n",
    "        args.min_acc = 0.6\n",
    "        args.mm2_ksize = 14\n",
    "        # args.alignment_threshold = 0.5\n",
    "    if args.isoseq:\n",
    "        args.min_mem = 20\n",
    "        args.min_acc = 0.8\n",
    "        # args.alignment_threshold = 0.5\n",
    "    return args"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "866cb84c-0f0a-48e1-abfd-ec369fdc045e",
   "metadata": {},
   "outputs": [],
   "source": [
    "def set_desalt_args(args):\n",
    "    d = 10 \n",
    "    s = 2\n",
    "    l = 14\n",
    "    noncan = 9\n",
    "    max_intron_length = 500000\n",
    "    index_path = args.desalt_index\n",
    "    \n",
    "    if(args.gtf == None):\n",
    "        if(\"SIRV\" in args.ref):\n",
    "            noncan = 4\n",
    "            max_intron_length = 200000\n",
    "            return [\"desalt\", \"aln\", index_path, args.reads, \n",
    "                               \"-d\", str(d), \"-s\", str(s),\n",
    "                                \"--noncan\", str(noncan), \"--max-intron-len\", str(max_intron_length),\n",
    "                    \"-o\", os.path.join(args.outfolder, \"reads.sam\")]\n",
    "        else:\n",
    "            return [\"desalt\", \"aln\", index_path, args.reads, \n",
    "                               \"-d\", str(d), \"-s\", str(s),\n",
    "                                \"--noncan\", str(noncan), \"--max-intron-len\", str(max_intron_length),\n",
    "                    \"-o\", os.path.join(args.outfolder, \"reads.sam\")]\n",
    "    else:\n",
    "        if(\"SIRV\" in args.ref):\n",
    "            noncan = 4\n",
    "            max_intron_length = 200000\n",
    "            return [\"desalt\", \"aln\", index_path, args.reads, \n",
    "                               \"-d\", str(d), \"-s\", str(s),\n",
    "                                \"--noncan\", str(noncan), \"--max-intron-len\", str(max_intron_length),\n",
    "                                \"--gtf\", args.gtf,\n",
    "                                \"-o\", os.path.join(args.outfolder, \"reads.sam\")]\n",
    "        else:\n",
    "            return [\"desalt\", \"aln\", index_path, args.reads, \n",
    "                               \"-d\", str(d), \"-s\", str(s),\n",
    "                                \"--noncan\", str(noncan), \"--max-intron-len\", str(max_intron_length),\n",
    "                                \"--gtf\", args.gtf,\n",
    "                                \"-o\", os.path.join(args.outfolder, \"reads.sam\")]\n",
    "\n",
    "\n",
    "def deSALT(args):\n",
    "    reads = os.path.join(args.outfolder, \"reads.sam\")\n",
    "    if (not os.path.exists(reads)):\n",
    "        stats = {}\n",
    "        stats[\"dataset\"] = args.name\n",
    "        initialize_dump(args.outfolder)\n",
    "        deSALT_start = time()\n",
    "        if not os.path.exists(args.desalt_index): \n",
    "            subprocess.check_call([\"desalt\", \"index\",args.ref, os.path.join(args.outfolder, \"index\")])\n",
    "        subprocess.check_call(set_desalt_args(args), env = os.environ)\n",
    "        stats[\"total_time\"] = time() - deSALT_start\n",
    "        json.dump(stats, open(os.path.join(args.outfolder, \"stats.json\"), \"w\"))\n",
    "\n",
    "\n",
    "    subprocess.check_call([\"desalt\", \"aln\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "4e85fc44-0144-462d-9f20-a1aef6a96cc2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create a dict with the paths of each dataset\n",
    "dataset_dict = {\n",
    "    # \"test\" : {\n",
    "    #     \"ref\": \"uLTRA/test/SIRV_genes.fasta\" ,\n",
    "    #     \"reads\": \"uLTRA/test/reads.fa\" ,\n",
    "    #     \"gtf\": \"uLTRA/test/SIRV_genes_C_170612a.gtf\" ,\n",
    "    #     \"bed\":\"uLTRA/test/SIRV_genes_C_170612a.bed\",\n",
    "    #     \"ont\": False,\n",
    "    #     \"isoseq\": True,\n",
    "    #     \"disable_infer\": False,\n",
    "    # },\n",
    "     \"DROS\": {\"ref\": \"data/genome/DROS/DROS.BDGP6.28.all.fa\",\n",
    "            \"reads\": \"data/DROS/DROS_processed.fastq\",\n",
    "            \"gtf\": \"data/genome/annotations/Drosophila_melanogaster.BDGP6.28.102.gtf\",\n",
    "            \"bed\":\"data/genome/annotations/bed/Drosophila_melanogaster.BDGP6.28.102.bed\",\n",
    "            \"desalt_index\":\"data/genome/indexes/dmIndex\",\n",
    "            \"ont\": True,\n",
    "            \"isoseq\": False,\n",
    "            \"disable_infer\": True,\n",
    "            \"desalt_index\":\"data/genome/indexes/dm\",   \n",
    "    },\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "dc1bc98e-8d51-49de-aae8-eac2ea7d5873",
   "metadata": {},
   "outputs": [],
   "source": [
    "def real_data_pipeline(dataset):\n",
    "    category_stats = {\"uLTRA\" :{}, \"uLTRA_mm2\" :{}, \"minimap2\" :{}, \"minimap2_GTF\" :{}, \"deSALT\" :{}, \"deSALT_GTF\" :{}}\n",
    "    \n",
    "    ref = dataset[\"ref\"]\n",
    "    gtf = dataset[\"gtf\"]\n",
    "    bed = dataset[\"bed\"]\n",
    "    reads = dataset[\"reads\"]\n",
    "    ont = dataset[\"ont\"]\n",
    "    isoseq = dataset[\"isoseq\"]\n",
    "    disable_infer = dataset[\"disable_infer\"]\n",
    "    desalt_index = dataset[\"desalt_index\"]\n",
    "    \n",
    "    # ultra\n",
    "    toolname = \"deSALT\"\n",
    "    args = create_args(ref, None, bed, reads, ont, isoseq, disable_infer, dataset_name, toolname, True, desalt_index)\n",
    "    deSALT(args)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "05291cb6-3bdd-47b0-89a7-ae6f5973769c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "creating output/DROS/deSALT\n",
      "File error:wrong route or wrong file name\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[Main] deSALT - De Bruijn graph-based Spliced Aligner for Long Transcriptome reads\n",
      "[Param-INFO] deSALT parameters:index-kmer:22\tseed-lmer:15\thash-kmer:8\tthread:4\tstrand_diff:10\tidentify junction:both_strand\n",
      "[Phase-INFO] Loading Index and Reads\n"
     ]
    },
    {
     "ename": "CalledProcessError",
     "evalue": "Command '['desalt', 'aln', 'output/DROS/deSALT/index', 'data/DROS/DROS_processed.fastq', '-d', '10', '-s', '2', '--noncan', '9', '--max-intron-len', '500000', '-o', 'output/DROS/deSALT/reads.sam']' returned non-zero exit status 1.",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mCalledProcessError\u001b[0m                        Traceback (most recent call last)",
      "Input \u001b[0;32mIn [11]\u001b[0m, in \u001b[0;36m<cell line: 1>\u001b[0;34m()\u001b[0m\n\u001b[1;32m      3\u001b[0m is_real \u001b[38;5;241m=\u001b[39m dataset[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mont\u001b[39m\u001b[38;5;124m\"\u001b[39m] \u001b[38;5;129;01mor\u001b[39;00m dataset[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124misoseq\u001b[39m\u001b[38;5;124m\"\u001b[39m] \u001b[38;5;66;03m# if the dataset is real\u001b[39;00m\n\u001b[1;32m      4\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m is_real:\n\u001b[0;32m----> 5\u001b[0m     \u001b[43mreal_data_pipeline\u001b[49m\u001b[43m(\u001b[49m\u001b[43mdataset\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m      6\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m      7\u001b[0m     sim_data_pipeline(dataset)\n",
      "Input \u001b[0;32mIn [10]\u001b[0m, in \u001b[0;36mreal_data_pipeline\u001b[0;34m(dataset)\u001b[0m\n\u001b[1;32m     14\u001b[0m toolname \u001b[38;5;241m=\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mdeSALT\u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m     15\u001b[0m args \u001b[38;5;241m=\u001b[39m create_args(ref, \u001b[38;5;28;01mNone\u001b[39;00m, bed, reads, ont, isoseq, disable_infer, dataset_name, toolname, \u001b[38;5;28;01mTrue\u001b[39;00m, desalt_index)\n\u001b[0;32m---> 16\u001b[0m \u001b[43mdeSALT\u001b[49m\u001b[43m(\u001b[49m\u001b[43margs\u001b[49m\u001b[43m)\u001b[49m\n",
      "Input \u001b[0;32mIn [8]\u001b[0m, in \u001b[0;36mdeSALT\u001b[0;34m(args)\u001b[0m\n\u001b[1;32m     46\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m os\u001b[38;5;241m.\u001b[39mpath\u001b[38;5;241m.\u001b[39mexists(args\u001b[38;5;241m.\u001b[39mdesalt_index): \n\u001b[1;32m     47\u001b[0m     subprocess\u001b[38;5;241m.\u001b[39mcheck_call([\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mdesalt\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mindex\u001b[39m\u001b[38;5;124m\"\u001b[39m,args\u001b[38;5;241m.\u001b[39mref, os\u001b[38;5;241m.\u001b[39mpath\u001b[38;5;241m.\u001b[39mjoin(args\u001b[38;5;241m.\u001b[39moutfolder, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mindex\u001b[39m\u001b[38;5;124m\"\u001b[39m)])\n\u001b[0;32m---> 48\u001b[0m \u001b[43msubprocess\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mcheck_call\u001b[49m\u001b[43m(\u001b[49m\u001b[43mset_desalt_args\u001b[49m\u001b[43m(\u001b[49m\u001b[43margs\u001b[49m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43menv\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43m \u001b[49m\u001b[43mos\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43menviron\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m     49\u001b[0m stats[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mtotal_time\u001b[39m\u001b[38;5;124m\"\u001b[39m] \u001b[38;5;241m=\u001b[39m time() \u001b[38;5;241m-\u001b[39m deSALT_start\n\u001b[1;32m     50\u001b[0m json\u001b[38;5;241m.\u001b[39mdump(stats, \u001b[38;5;28mopen\u001b[39m(os\u001b[38;5;241m.\u001b[39mpath\u001b[38;5;241m.\u001b[39mjoin(args\u001b[38;5;241m.\u001b[39moutfolder, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mstats.json\u001b[39m\u001b[38;5;124m\"\u001b[39m), \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mw\u001b[39m\u001b[38;5;124m\"\u001b[39m))\n",
      "File \u001b[0;32m~/opt/anaconda3/envs/ultra/lib/python3.8/subprocess.py:364\u001b[0m, in \u001b[0;36mcheck_call\u001b[0;34m(*popenargs, **kwargs)\u001b[0m\n\u001b[1;32m    362\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m cmd \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[1;32m    363\u001b[0m         cmd \u001b[38;5;241m=\u001b[39m popenargs[\u001b[38;5;241m0\u001b[39m]\n\u001b[0;32m--> 364\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m CalledProcessError(retcode, cmd)\n\u001b[1;32m    365\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;241m0\u001b[39m\n",
      "\u001b[0;31mCalledProcessError\u001b[0m: Command '['desalt', 'aln', 'output/DROS/deSALT/index', 'data/DROS/DROS_processed.fastq', '-d', '10', '-s', '2', '--noncan', '9', '--max-intron-len', '500000', '-o', 'output/DROS/deSALT/reads.sam']' returned non-zero exit status 1."
     ]
    }
   ],
   "source": [
    "for dataset_name in dataset_dict.keys():\n",
    "    dataset = dataset_dict[dataset_name]\n",
    "    is_real = dataset[\"ont\"] or dataset[\"isoseq\"] # if the dataset is real\n",
    "    if is_real:\n",
    "        real_data_pipeline(dataset)\n",
    "    else:\n",
    "        sim_data_pipeline(dataset)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5c0f3f5e-a53a-4fd2-ac64-7d04fabaf924",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
